src/components/users/users.ts

```diff
@@ -163,31 +179,39 @@ export async function listUsers(ctx: IContext, params: IParameters): Promise<IRe
     CLOUD_CONTROLLER_GLOBAL_AUDITOR,
   );
 
-  const [isManager, isBillingManager, organization, users] = await Promise.all([
+  const [isManager, isBillingManager, organization, userOrgRoles] = await Promise.all([
     cf.hasOrganizationRole(params.organizationGUID, ctx.token.userID, 'org_manager'),
     cf.hasOrganizationRole(params.organizationGUID, ctx.token.userID, 'billing_manager'),
     cf.organization(params.organizationGUID),
     cf.usersForOrganization(params.organizationGUID),
   ]);
 
-  const usersWithSpaces = await Promise.all(users.map(async (user: IOrganizationUserRoles) => {
-    const userWithSpaces = {
-      ...user,
-      spaces: new Array(),
-    };
+  const spacesVisibleToUser = await cf.spaces(params.organizationGUID);
 
-    /* istanbul ignore next */
-    try {
-      userWithSpaces.spaces = [...await cf.spacesForUserInOrganization(user.metadata.guid, params.organizationGUID)];
-    } catch {
-      ctx.log.warn(
-        `BUG: users has no permission to fetch spacesForUser: ${user.metadata.guid}`,
-      ); // TODO: permissions issue here
-    }
+  const usersOrgRolesByGuid: IUserByGuid =
+    userOrgRoles.reduce((acc, user) => ({...acc, [user.metadata.guid]: user}), {});
 
-    return userWithSpaces;
+  const spaceUserLists = await Promise.all(spacesVisibleToUser.map(async space => {
+    return {space, users: await cf.usersForSpace(space.metadata.guid)};
   }));
 
+  const userRolesByGuid: IUserRolesByGuid = {};
+
+  for (const spaceUsers of spaceUserLists) {
```

Comment:
I've been thinking about it a bit more, and I think we should pull this big complicated bit out into a function which can be tested independently.

One bug I can think of at the moment is: if a user is in the org, but isn't in any spaces that the logged in user can see, they won't be able to see their org roles (because they're only assigned inside the spaces loop).

Reply:
> I've been thinking about it a bit more, and I think we should pull this big complicated bit out into a function which can be tested independently.

Agreed. I'll take a look at that

> if a user is in the org, but isn't in any spaces that the logged in user can see, they won't be able to see their org roles

Actually, it seems worse than that. I'll move this PR back to WIP